### ISSUE TITLE
libssh2_session_handshake returns -44  with OpenSSL crypto backend

### ISSUE THREAD

**Issue by SX-80** [libssh2/libssh2#463]
The function libssh2_session_handshake returns -44 "Unable to ask for ssh-userauth service" in code using OpenSSL crypto backend.

```
    // ... initialize socket ...

    LIBSSH2_SESSION* session = libssh2_session_init();
    int res = libssh2_session_handshake( session, socket );

    char const* str = nullptr;
    int len = 0;
    libssh2_session_last_error( session, &str, &len, false );

    std::cout << "res: " << res << ", error: " << str << std::endl;
```

Error is found in function ''_libssh2_cipher_crypt" (openssl.c: 416)
The problem appears after commit in OpenSSL:
https://github.com/openssl/openssl/commit/f7397f0d58ce7ddf4c5366cd1846f16b341fbe43

**Version**
 - OS: Windows 7
 - libssh2 version [1.9.0_DEV]              - master on 2020-04-27
 - openssl version [3.0.0-alpha2-dev]  - master on 2020-04-27

**Comment by willco007** [comment]
Have you tried a shipping version of OpenSSL? Seems like a non-zero chance this is a bug in the alpha release of OpenSSL you're using.

**Comment by SX-80** [comment]
Corrected version
```
int
_libssh2_cipher_crypt(_libssh2_cipher_ctx * ctx,
                      _libssh2_cipher_type(algo),
                      int encrypt, unsigned char *block, size_t blocksize)
{
    unsigned char buf[EVP_MAX_BLOCK_LENGTH];
    int ret;
    (void) algo;
    (void) encrypt;

#ifdef HAVE_OPAQUE_STRUCTS
    ret = EVP_Cipher(*ctx, buf, block, blocksize);
#else
    ret = EVP_Cipher(ctx, buf, block, blocksize);
#endif
    if(0 < ret) {
        memcpy(block, buf, blocksize);
    }
    return 0 < ret ? 0 : 1;
}

**Comment by SX-80** [comment]
The changes in OpenSSL library
![изображение](https://user-images.githubusercontent.com/64437815/80421138-a9264500-88e4-11ea-9b74-177eeaeee56e.png)


Now returns 'outl'  or '-1'

**Comment by SX-80** [comment]
![изображение](https://user-images.githubusercontent.com/64437815/80421378-0cb07280-88e5-11ea-92fb-ad9e735db51f.png)

**Comment by willco007** [comment]
Thanks for tracking this down, I'll drop in a fix.

**Comment by willco007** [comment]
Looking at the git commit for OpenSSL it seems like ret <= ret is success, not just 0 < ret.

**Comment by willco007** [comment]
Furthermore, there is this mentioned in their commit:

`EVP_Cipher() returns the amount of encrypted / decrypted bytes, or -1
 on failure, if the flag B<EVP_CIPH_FLAG_CUSTOM_CIPHER> is set for the
 cipher.  EVP_Cipher() returns 1 on success or 0 on failure, if the flag
 B<EVP_CIPH_FLAG_CUSTOM_CIPHER> is not set for the cipher.
`

To my knowledge we're not setting this flag so it should be returning 1, not bytes encrypted.

**Comment by willco007** [comment]
Looks like this is set by the cipher inside of OpenSSL implicitly. Their release notes: `New flag in ciphers: EVP_CIPH_FLAG_CUSTOM_CIPHER. This means the underlying do_cipher function handles all cipher semantics itself including padding and finalisation. ` Our API call is going to have to check which version of OpenSSL libssh2 is using to handle the possible return code values since 1.1/1.0/3.0 aren't compatible with each other.

**Comment by SX-80** [comment]
> Looking at the git commit for OpenSSL it seems like 0 <= ret is success, not just 0 < ret.



You are right, but "0 < ret" decision works with both version of OpenSSL with no other code.
Correct me if I am wrong, but I expect that libssh2 not calls _libssh2_cipher_crypt with blockssize == 0.
This is one of the simplest solution.



> Our API call is going to have to check which version of OpenSSL libssh2
Of course the correct solution is to CHECK MACRO and use special code for every possible case

**Comment by SX-80** [comment]
> To my knowledge we're not setting this flag so it should be returning 1

I am not setting this flag too.  But function EVP_Cipher in my application calls ctx->cipher->ccipher.

![изображение](https://user-images.githubusercontent.com/64437815/80449264-54f08480-8927-11ea-8e2e-66fccf0b733d.png)

**Comment by SX-80** [comment]
> but "0 < ret" decision works with both versions of OpenSSL with no other code

and always works whether the function calls   ctx->cipher->ccipher,   ctx->cipher->cupdate  or  ctx->cipher->cfinal.
(assumption: libssh2 not calls _libssh2_cipher_crypt with blockssize == 0)

**Comment by jmarchil** [comment]
Is there any follow-up on this issue. I am integrating with openssl 3.0 alpha 11 and I am seeing this issue.

**Comment by willco007** [comment]
A fix was pushed with #463; could all you who are running into this verify this is fixed?

**Comment by jmarchil** [comment]
> 
> 
> A fix was pushed with #463; could all you who are running into this verify this is fixed?

I verified using the latest openssl 3.0 alpha 17 and the problem is resolved

**Comment by willco007** [comment]
@jmarchil Great thanks! I'm going to close this issue.
